Libraries/oneMKL/fourier_correlation/fcorr_2d_usm.cpp

```diff
@@ -6,151 +6,291 @@
+  // normalization purposes and for the definition of error tolerance)
+  const float norm_img1 =
+          calc_frobenius_norm(img1, Q, n_rows, n_cols, col_stride_fwd_domain);
+  const float norm_img2 =
+          calc_frobenius_norm(img2, Q, n_rows, n_cols, col_stride_fwd_domain);
+  // 1) Calculate the cross-correlation naively (for verification purposes);
+  naive_cross_correlation(Q, n_rows, n_cols, col_stride_fwd_domain,
+                          img1, img2, naive_corr).wait();
+  // 2) Calculate the cross-correlation via Discrete Fourier Transforms (DFTs):
+  //         corr = (1/num_elem) * iDFT(DFT(sig1) * CONJ(DFT(sig2)))
+  // Initialize DFT descriptor
+  oneapi::mkl::dft::descriptor<oneapi::mkl::dft::precision::SINGLE,
+                               oneapi::mkl::dft::domain::REAL>
+    desc({n_rows, n_cols});
+  desc.set_value(oneapi::mkl::dft::config_param::BACKWARD_SCALE,
+                 1.0f / num_elem);
+  desc.commit(Q);
+  // --> desc operates in-place with the strides used herein by default.
 
-    backward_plan.commit(Q);
-    auto bwd = oneapi::mkl::dft::compute_backward(backward_plan, corr);
-    bwd.wait();
+  // Compute in-place forward transforms of both signals:
+  // img1 <- DFT(img1)
+  auto evt1 = oneapi::mkl::dft::compute_forward(desc, img1);
+  // img2 <- DFT(img2)
+  auto evt2 = oneapi::mkl::dft::compute_forward(desc, img2);
+  // Compute the element-wise multipication of (complex) coefficients in
+  // backward domain:
+  // corr <- sig1 * CONJ(sig2) [component-wise]
+  evt = oneapi::mkl::vm::mulbyconj(
+          Q, n_rows * (n_cols / 2 + 1),
+          reinterpret_cast<std::complex<float>*>(img1),
+          reinterpret_cast<std::complex<float>*>(img2),
+          reinterpret_cast<std::complex<float>*>(corr),
+          {evt1, evt2});
+  // Compute in-place (scaled) backward transform:
+  // corr <- (1 / num_elem) * iDFT(corr)
+  oneapi::mkl::dft::compute_backward(desc, corr, {evt}).wait();
 
-    std::cout << std::endl << "Normalized Fourier correlation result:"
-              << std::endl;
-    for (unsigned int r = 0; r < n_rows; r++) {
-        for (unsigned int c = 0; c < n_cols; c++) {
-            std::cout << corr[r * c_stride + c * r_stride] / (n_rows * n_cols)
-                      << " ";
-        }
-        std::cout << std::endl;
+  // Error bound for naive calculations:
+  float max_err_threshold =
+    2.0f * std::numeric_limits<float>::epsilon() * norm_img1 * norm_img2;
+  // Adding an (empirical) error bound for the DFT-based calculation defined as
+  //    epsilon * O(log(num_elem)) * scaling_factor * nrm2(input data),
+  // wherein (for the last DFT at play)
+  // - scaling_factor = 1.0 / num_elem;
+  // - nrm2(input data) = norm_sig1[0] * norm_sig2[0] * num_elem
+  // - O(log(num_elem)) ~ 2 * log(num_elem)
+  //   [arbitrary choice; implementation-dependent behavior]
+  max_err_threshold +=
+    2.0f * logf(num_elem) * std::numeric_limits<float>::epsilon()
```

Comment:
Do you mean `logf` from C, `std::logf`, sycl::logf? I'd assume it's `std::logf`, right? Could you please explicitly specify it here and for other math functions below?

P.S. The same comment is for 2 source files above

Reply:
Changed and unified to `std::{log, sqrt}(float)` in 7e7073fb6dd141512dc3531086b5b2539f9cb803